LwM2M: fix block transfer with TLV

jb-gcx opened the pull request:
Follow-up packets do not contain the TLV header but just continuations of the value. Tested with a Leshan server and a custom opaque resource.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/34935

Signed-off-by: Jan Bünker <jan.buenker@grandcentrix.net>

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c:
```diff
@@ -935,23 +935,50 @@ static int do_write_op_tlv_item(struct lwm2m_message *msg)
+
+	msg->path.res_id = tlv->id;
+	msg->path.level = 3U;
+	int ret = do_write_op_tlv_item(msg);
```
Please move the variable declaration to the beginning of the block.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c:
```diff
@@ -935,23 +935,50 @@ static int do_write_op_tlv_item(struct lwm2m_message *msg)
+	 * optional resources are ignored (ENOTSUP)
+	 */
+	if (ret < 0 &&
+		!((ret == -ENOTSUP) &&
```
Would be nice to keep the indentation from the original code.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c:
```diff
@@ -995,36 +1022,14 @@ int do_write_op_tlv(struct lwm2m_message *msg)
-				    !((ret == -ENOTSUP) &&
-				      (msg->ctx->bootstrap_mode ||
-				       msg->operation == LWM2M_OP_CREATE))) {
+				if ((ret = write_tlv_resource(msg, &tlv2))) {
```
We typically avoid assignments inside the `if` condition
```suggestion
				ret = write_tlv_resource(msg, &tlv2);
				if (ret < 0) {
```

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c:
```diff
@@ -995,36 +1022,14 @@ int do_write_op_tlv(struct lwm2m_message *msg)
-			    !((ret == -ENOTSUP) &&
-			      (msg->ctx->bootstrap_mode ||
-			       msg->operation == LWM2M_OP_CREATE))) {
+			if ((ret = write_tlv_resource(msg, &tlv))) {
```
Same here.

rlubos reviewed (commented):
Thanks, current implementation indeed was limited to firmware update only, thanks for making this more generic. It also leaves an opening for further improvements in the future, as currently block transfer writes are limited to a single resource only.

You'd need to update your commit message in order to satisfy Zephyr CI - please add `net: lwm2m: ` prefix to the commit title, a signoff entry and stick to char/line limit of the commit message. You can check what's wrong in the compliance check results.

rlubos reviewed (approved):
One more nit, there's a duplicated signoff entry in the commit message, but otherwise looks good.

jb-gcx commented:
Thanks for the Feedback @rlubos ! After editing the existing commit message, git appended the signoff message a second time and I didn't notice. I hope everything is in order now.

jukkar reviewed (approved):
LGTM

rlubos commented:
@jukkar Any idea why is the gitlint complaining about the signoff entries? It looks correct to me...

rlubos commented:
> The original message had the signoff entry as well, and it also complained. I figured gitlint might have gotten confused because of other issues with the message, but those are fixed now.
> 
> Maybe it doesn't like the umlaut?

I'm not sure to be honest - it doesn't seem like any of the typical issues I've seen so far. gitlint is fairly easy to run locally, perhaps you could give it a show w/o umlaut. Just not that another gitlint requirement is that the commit author name matches the signoff entry, so you might need to run `--reset-author` option on the commit as well if you modify it.

jb-gcx commented:
I saw the issue locally, amended the commit without any changes and without `-s`, then re-ran gitlint and it worked. If this works, then the cause seems to be gits automatic signoff feature, although I don't understand why.

rlubos commented:
@jb-gcx Now the complaint is different:
> author email (Jan Bünker <jan.buenker@grandcentrix.net>) needs to match one of the signed-off-by entries.

jb-gcx commented:
I'm stumped by this one. My local gitlint is OK with everything, and the `git log` output looks correct as well:
```
commit 55e6c527b312eafc28f467058906e93035a4ba76 (HEAD -> fix_lwm2m_block_transfer_tlv, jb-gcx-fork/fix_lwm2m_block_transfer_tlv)
Author: Jan Bünker <jan.buenker@grandcentrix.net>
Date:   Tue May 4 16:43:49 2021 +0200

    net: lwm2m: Only parse TLV from the first block
    
    This was already implemented for firmware update packages.
    For other opaque resources it failed to determine the target resource
    id, which is now stored in the block_context.
    
    Signed-off-by: Jan Bünker <jan.buenker@grandcentrix.net>
```

Curious: the output of the online check references `b6b1733d7e895f227ce255d1ee73caf9b222c720` - if that's supposed to be a commit hash, it's the wrong one.

rlubos commented:
> Curious: the output of the online check references `b6b1733d7e895f227ce255d1ee73caf9b222c720` - if that's supposed to be a commit hash, it's the wrong one.

It's likely due to the fact, that compliance check rebases the PR branch to the latest master before running the checks - hence different SHA of your commit. 

I'm not gitlint expert, but from experience I can tell that it only reports results if the `gitlint` command was issued from the root directory of the repository - perhaps you did otherwise and that's why you can't see the error locally?

jb-gcx commented:
> I'm not gitlint expert, but from experience I can tell that it only reports results if the `gitlint` command was issued from the root directory of the repository - perhaps you did otherwise and that's why you can't see the error locally?

Unfortunately no, I'm in the root directory. I've tried rebasing onto master and pushing again, maybe this will help. I thought maybe the rebase modifies the author somehow. Although then this error would occur a lot, so it doesn't seem likely.

jb-gcx commented:
Maybe there is a difference in behavior between linux and macos? I am on mac OS.

carlescufi commented:
It's an encoding problem:

Your email: `b'Jan Bu\xcc\x88nker <jan.buenker@grandcentrix.net>'`
Your signed-off-by line: `b'Jan B\xc3\xbcnker <jan.buenker@grandcentrix.net>'`

The first one is using an actual ASCII `u` character and then [this modifier](https://unicode.scarfboy.com/?s=0308) and the second is using [this one character](https://unicode.scarfboy.com/?s=U%2B00FC)

jb-gcx commented:
> It's an encoding problem:
> 
> Your email: `b'Jan Bu\xcc\x88nker <jan.buenker@grandcentrix.net>'`
> Your signed-off-by line: `b'Jan B\xc3\xbcnker <jan.buenker@grandcentrix.net>'`
> 
> The first one is using an actual ASCII `u` character and then [this modifier](https://unicode.scarfboy.com/?s=0308) and the second is using [this one character](https://unicode.scarfboy.com/?s=U%2B00FC)

Thanks, I didn't see that!

I am now using the same encoding as the author field in the commit message, but the old error is back: `1: UC2 Signed-off-by: must have a full name`. I guess it doesn't like that encoding.

But now I know why it worked one time, when I had the double signoff message; the two lines must have been encoded differently, so gitlint used one encoding to compare to the author and the other one to check for a full name.

I'm out of time for today, but tomorrow I'll either change the encoding of my author field, or just use "Buenker" everywhere to make myself future proof.

jb-gcx committed:
net: lwm2m: Only parse TLV from the first block

This was already implemented for firmware update packages.
For other opaque resources it failed to determine the target resource
id, which is now stored in the block_context.

Signed-off-by: Jan Buenker <jan.buenker@grandcentrix.net>

ziposoft-ac commented on subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c:
```diff
@@ -935,23 +935,53 @@ static int do_write_op_tlv_item(struct lwm2m_message *msg)
 			return ret;
 		}
+
+		return 0;
```
This "return 0;" seems to cause a problem.
If the current packet contains writes to other resources in the object, those will not be processed. Only the first item is written, and function returns without processing the rest.
If I remove the "return 0", it seems to resolve my problem and the other writes are processed. However, that "return 0" was specifically added for a reason, so I am not confident that is the correct fix.
